### PR TITLE
Add some useful debug information to some exceptions

### PR DESCRIPTION
**Pull request by Mingun**

### CODE REVIEW

File: spec/unit/compiler/passes/helpers.js [hunk 1]
```diff
@@ -68,7 +68,7 @@ beforeEach(function() {
           this.message = () =>
             "Expected the pass not to report an error "
               + "for grammar " + jasmine.pp(grammar) + ", "
-              + "but it did.";
+              + "but it did with message " + jasmine.pp(e.message);
```
Comment (dmajda) [inline spec/unit/compiler/passes/helpers.js]: Please replace “but it did with message” with more grammatically correct “but it reported one with message”. Also, there was a comma at the end of the message.

File: lib/compiler/passes/generate-js.js [hunk 2]
```diff
@@ -383,7 +383,7 @@ function generateJS(ast, options) {
       "          break;",
       "",
       "        default:",
-      "          throw new Error(\"Invalid opcode: \" + bc[ip] + \".\");",
+      "          throw new Error(\"Invalid opcode #\" + ip + \": \" + bc[ip] + \".\");",
```
Comment (dmajda) [inline lib/compiler/passes/generate-js.js]: I’m curious: you apparently missed the opcode position in the message but you didn’t miss an indication in bytecode of *which rule* the invalid opcode was? I’d rather add both, e.g. “Invalid opcode: 77 (rule: 5, position: 6).”.

File: lib/compiler/passes/generate-js.js [hunk 3]
```diff
@@ -473,7 +473,7 @@ function generateJS(ast, options) {
 
           if (thenSp !== elseSp) {
             throw new Error(
-              "Branches of a condition must move the stack pointer in the same way."
+              "Branches of a condition must move the stack pointer in the same way (then=" + thenSp + ", else=" + elseSp + ")."
```
Comment (dmajda) [inline lib/compiler/passes/generate-js.js]: I’d format the message like this for more readability: “Branches of a condition must move the stack pointer in the same way (before: 5, after then: 6, after else: 4).”. Or maybe even like this: “Branches of a condition can't move the stack pointer differently (before: 5, after then: 6, after else: 4).”.

Also please split the concatenation over multiple lines.

File: lib/compiler/passes/generate-js.js [hunk 4]
```diff
@@ -499,7 +499,7 @@ function generateJS(ast, options) {
         ip += bodyLength;
 
         if (bodySp !== baseSp) {
-          throw new Error("Body of a loop can't move the stack pointer.");
+          throw new Error("Body of a loop can't move the stack pointer (base=" + baseSp + ", body=" + bodySp + ").");
```
Comment (dmajda) [inline lib/compiler/passes/generate-js.js]: I’d format the message like this for more readability: “Body of a loop can't move the stack pointer (before: 5, after: 6).”.

Also please split the concatenation over multiple lines.

File: lib/compiler/passes/generate-js.js [hunk 5]
```diff
@@ -705,7 +705,7 @@ function generateJS(ast, options) {
             break;
 
           default:
-            throw new Error("Invalid opcode: " + bc[ip] + ".");
+            throw new Error("Invalid opcode #" + ip + ": " + bc[ip] + ".");
```
Comment (dmajda) [inline lib/compiler/passes/generate-js.js]: Like above, I’d add both the position and the rule into the message: “Invalid opcode: 77 (rule: Expression, position: 6).”.

**Comment by dmajda** [comment]
As a side note, the Travis CI [build failure](https://travis-ci.org/pegjs/pegjs/builds/179220657) is unrelated to this change and I [fixed it](https://github.com/pegjs/pegjs/commit/a733beb933d67b725c4972c97b8e5929792224bd) yesterday. Rebasing on top of current master should result in a clean build.

**Comment by Mingun** [comment]
Any comments?

**Comment by Mingun** [comment]
@futagoza?

**Comment by Mingun** [comment]
Now this done and can be merged.

**Comment by futagoza** [comment]
I prefer `fn` over `func`, but oh well, doesn't matter 😄

**Pull request merged**